File: yht_custom/public/js/list_defaults.js
// Copyright (c) 2026, Enfono Technologies and contributors
// For license information, please see license.txt
//
// Client sheet item 35 — a transaction list opens on the current fiscal year.
//
// 🔴 LOADED FROM `doctype_list_js`, NOT `app_include_js`, AND THAT IS THE WHOLE
// TRICK. erpnext's own list JS opens with a WHOLESALE assignment to
// `frappe.listview_settings["X"]`, and a doctype's list bundle is fetched when
// the list is first opened — long after `app_include_js` has run. Anything
// merged in at boot is discarded silently: no error, the default simply never
// applies. `frappe/desk/form/meta.py` adds the doctype's own `<doctype>_list.js`
// first and THEN concatenates the `doctype_list_js` hook file, so a merge from
// here lands after erpnext's assignment and survives.
//
// Which is also why this file only ever MERGES. It never assigns that key.
//
// ⚠️ WRAPPED IN AN IIFE ON PURPOSE. One file is registered under eight
// doctypes and the desk evaluates each doctype's list JS in global scope, so a
// top-level `const` would throw "Identifier has already been declared" the
// second time a user opens one of these lists in a session.
(function () {
	// The eight doctypes that carry `custom_fiscal_year` — the same set as
	// `yht_custom.fiscal_year.DATE_FIELD`, which is what hooks.py registers this
	// file under.
	const LISTS = [
		"Sales Invoice",
		"Purchase Invoice",
		"Delivery Note",
		"Purchase Receipt",
		"Payment Entry",
		"Journal Entry",
		"Sales Order",
		"Quotation",
	];
	const FIELD = "custom_fiscal_year";

	// Resolved server-side in `boot.boot_session` because `setup_defaults()`
	// builds the list synchronously — there is no moment to fetch it in. An
	// empty value means no Fiscal Year covers today, and then no filter is set
	// at all rather than one naming a year that does not exist.
	if (!frappe.boot.yht_current_fiscal_year) return;

	frappe.provide("frappe.listview_settings");

	for (const doctype of LISTS) {
		frappe.provide(`frappe.listview_settings.${doctype}`);
		const settings = frappe.listview_settings[doctype];

		const existing = Array.isArray(settings.filters) ? settings.filters : [];
		// Appended, not replaced: a doctype's own list JS may already ship a
		// default filter, and dropping it here would be a silent regression.
		// Guarded because this file runs again every time one of the eight lists
		// is opened for the first time in a session.
		if (existing.some((f) => Array.isArray(f) && f.indexOf(FIELD) !== -1)) continue;

		settings.filters = existing.concat([[FIELD, "=", frappe.boot.yht_current_fiscal_year]]);

	}
})();

File: yht_custom/list_defaults.py
# Copyright (c) 2026, Enfono Technologies and contributors
# For license information, please see license.txt

"""What a transaction list opens on, before anyone touches a filter.

Client sheet item 35: the eight fiscal-year lists should open on the CURRENT
year rather than on everything ever imported.

**Mechanism**, verified in
`frappe/public/js/frappe/list/list_view.js::setup_defaults`:

    if (this.view_user_settings.filters && this.view_user_settings.filters.length) {
        this.filters = this.validate_filters(saved_filters);   // 1: the user's own
    } else {
        this.filters = (this.settings.filters || []).map(...); // 2: listview_settings
    }

So `frappe.listview_settings[dt].filters` is the supported pre-set, and the
user's own saved filters always win over it.

🔴 THIS IS A FILTER DEFAULT, NEVER A FIELD DEFAULT. `custom_fiscal_year` stays
`read_only` with no `default` of any kind — a fixed default on the FIELD is what
left 731 Sales Invoices stamped FY 2026 while dated from 2024-08-01 on the
client's previous system. `fiscal_year.set_fiscal_year` computes the stored
value from the posting date on every save and that does not change here.

`setup_defaults()` runs synchronously while the list is being built, so the year
has to be in the boot payload before the list exists — hence
`boot.boot_session` rather than a call from the list JS. Bootinfo is cached per
user, so a session held open across a year-end keeps the old value until the
cache clears; `bench clear-cache` is already in the deploy sequence.
"""

from frappe.utils import today

from yht_custom import fiscal_year


def current_fiscal_year() -> str:
	"""The Fiscal Year covering today, or `""` when none does.

	Reuses `fiscal_year.resolve`, which is what makes this roll into 2027 on its
	own and inherits the standing rule that a fiscal-year lookup must never be
	the reason something fails: no year covering today gives an empty string, and
	the list JS then sets no filter at all.
	"""
	return fiscal_year.resolve(today())
